### PR TITLE
Improve visor config and startup/shutdown logic.

### LINKED ISSUE #170: Use wait group to ensure all components shutdown before Visor.Close returns.
**Feature description**

We should wait until all go routines of components shutdown before Visor.Close returns.
This should ensure a clean shutdown and may guarantee a successful restart.

**Describe the solution you'd like**
Passing a `sync.WaitGroup` to all go routines.

### LINKED ISSUE #173: Improve visor configuration experience
**Issues**
- A feature for reducing the likelihood of having misspelled/wrong config values. [(link)](https://github.com/SkycoinProject/skywire-mainnet/issues/173#issuecomment-594441598)
- Add points to config methods to avoid golang default values. [(link)](https://github.com/SkycoinProject/skywire-mainnet/issues/173#issuecomment-594450919)
- Generate config READMEs/help. [(link)](https://github.com/SkycoinProject/skywire-mainnet/issues/173#issuecomment-595135072)
- We should have a standard for configuration of execs. [(link)](https://github.com/SkycoinProject/skywire-mainnet/issues/173#issuecomment-595141495)

### LINKED ISSUE #339: Improvements to visor.Config
**Feature description**

Currently config generation is split across multiple areas (depending on the use case). Specifically in the following two files:
- https://github.com/SkycoinProject/skywire-mainnet/blob/master/pkg/visor/config.go
- https://github.com/SkycoinProject/skywire-mainnet/blob/master/cmd/skywire-cli/commands/visor/gen-config.go

My proposal is this:
- Have a single function within `visor` module defined as `visor.DefaultConfig() *Config`.
- This can be directly as the definition for `skywire-cli visor gen-config`.
- For missing fields when `skywire-visor` is run, we can do:
    ```golang
    conf := DefaultConfig()
    err := json.NewDecoder(file).Decode(conf)
    ```
    This way, we we set default values first, and user-set values will overwrite default values.

Further considerations:
- Maybe set `.DisallowUnknownFields()` for json decoder.

### PR DESCRIPTION
**Pull request by evanlinjin**
Fixes #339 fixes #173 fixes #170

 Changes:	
- Separated app-associated functionality within `/pkg/visor` into `/pkg/app/launcher`.
- Added the concept of `visor.BaseConfig()` to simplify obtaining default values.
- Made startup/shutdown logic modular and introduced init and shutdown stacks.
- Shutdown now uses module-based timeouts and wait groups.

### CODE REVIEW

File: cmd/skywire-cli/commands/visor/gen-config.go [hunk 1]
```diff
@@ -56,13 +53,13 @@ var genConfigCmd = &cobra.Command{
...
+
 		switch configLocType {
 		case pathutil.WorkingDirLoc:
-			conf = defaultConfig()
-		case pathutil.HomeLoc:
-			conf = homeConfig()
-		case pathutil.LocalLoc:
-			conf = localConfig()
+			var err error
+			if conf, err = visor.DefaultConfig(nil, output, visor.NewKeyPair()); err != nil {
```
Comment (Darkren) [inline cmd/skywire-cli/commands/visor/gen-config.go:62]: Do we really need `configLocType` anymore? seems like it should be removed entirely along with this switch

File: pkg/app/launcher/launcher.go [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+
+func (l *Launcher) startApp(cmd string, args, envs []string) error {
+	log := l.log.WithField("func", "StartApp").WithField("cmd", cmd)
+
+	// Obtain associated app config.
+	ac, ok := l.apps[cmd]
+	if !ok {
+		return ErrAppNotFound
+	}
+	if args != nil {
```
Comment (Darkren) [inline pkg/app/launcher/launcher.go:196]: We should be checking `len(args) != 0` imo here. Just to exclude non-nil but still empty args

File: pkg/app/launcher/launcher.go [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+	}
+	return nil
+}
+
+/*
+	<<< PID management >>>
+*/
+
+func (l *Launcher) pidFile() (*os.File, error) {
+	return os.OpenFile(filepath.Join(l.conf.LocalPath, "apps-pid.txt"), os.O_RDWR|os.O_CREATE, 0600)
```
Comment (Darkren) [inline pkg/app/launcher/launcher.go]: I know it's probably an old part, but could you please move the file name to a constant?

File: pkg/hypervisor/config.go [hunk 4]
```diff
@@ -48,7 +48,7 @@ func (hk *Key) UnmarshalText(text []byte) error {
 type Config struct {
 	PK            cipher.PubKey `json:"public_key"`
 	SK            cipher.SecKey `json:"secret_key"`
-	DBPath        string        `json:"db_path"`        // Path to store database file.
+	DBPath        string        `json:"db_path"`        // path to store database file.
```
Comment (Darkren) [inline pkg/hypervisor/config.go]: Is this accidental?

File: pkg/visor/config.go [hunk 5]
```diff
@@ -47,258 +38,270 @@ var (
...
+	}
+	if keys != nil {
+		conf.KeyPair = keys
+		conf.ensureKeys()
+	}
+	return conf, conf.Flush()
+}
+
+// UpdateAppAutostart modifies a single app's autostart value within the config and also the given launcher.
+// The updated config gets flushed to file is there are any changes.
```
Comment (Darkren) [inline pkg/visor/config.go]: `is` -> `if`, just a small mistyp

File: pkg/visor/config.go [hunk 6]
```diff
@@ -47,258 +38,270 @@ var (
...
-// If TransportConfig is not found, DefaultTransportConfig() is used.
-func (c *Config) TransportDiscovery() (transport.DiscoveryClient, error) {
-	if c.Transport == nil {
-		c.Transport = DefaultTransportConfig()
-		if err := c.flush(); err != nil && c.log != nil {
-			c.log.WithError(err).Errorf("Failed to flush config to disk")
-		}
-	}
+// UpdateAppArg updates the cli flag of the specified app config and also within the launcher.
+// The updated config gets flushed to file is there are any changes.
```
Comment (Darkren) [inline pkg/visor/config.go]: Same mistype as above

File: pkg/visor/visor_init.go [hunk 7]
```diff
@@ -0,0 +1,418 @@
+package visor
```
Comment (Darkren) [inline pkg/visor/visor_init.go]: Wouldn't `init` be enough for the file name? it's in the `visor` package anyway

**Review by Darkren — CHANGES_REQUESTED** [review]
Really huge improvement. Just some minor queries here and there

File: pkg/hypervisor/config.go (hunk 4 above)
Comment (evanlinjin) [reply pkg/hypervisor/config.go]: Yes, woops.

File: cmd/skywire-cli/commands/visor/gen-config.go (hunk 1 above)
Comment (evanlinjin) [reply cmd/skywire-cli/commands/visor/gen-config.go:62]: @Darkren I've left it in there for now as I think we should consider whether to re-implement this feature in the future.

Comment (evanlinjin) [reply cmd/skywire-cli/commands/visor/gen-config.go:62]: I'll leave a comment there for now.

File: pkg/app/launcher/launcher.go (hunk 2 above)
Comment (evanlinjin) [reply pkg/app/launcher/launcher.go:196]: @Darkren But what if people want to run an app with no args?

File: cmd/skywire-cli/commands/visor/app.go [hunk 8]
```diff
@@ -8,10 +8,11 @@ import (
 	"text/tabwriter"
 	"time"
 
+	"github.com/SkycoinProject/skywire-mainnet/pkg/app/launcher"
```
Comment (nkryuchkov) [inline cmd/skywire-cli/commands/visor/app.go]: Could you please sort imports?

File: pkg/app/appnet/skywire_networker.go [hunk 9]
```diff
@@ -9,8 +9,9 @@ import (
 	"sync"
 	"sync/atomic"
 
+	"github.com/sirupsen/logrus"
+
```
Comment (nkryuchkov) [inline pkg/app/appnet/skywire_networker.go]: Redundant new line

File: pkg/hypervisor/hypervisor.go [hunk 10]
```diff
@@ -13,6 +13,8 @@ import (
 	"sync"
 	"time"
 
+	"github.com/SkycoinProject/skywire-mainnet/pkg/app/launcher"
```
Comment (nkryuchkov) [inline pkg/hypervisor/hypervisor.go]: Please format imports

File: pkg/visor/config.go [hunk 11]
```diff
@@ -5,39 +5,30 @@ import (
...
 	"fmt"
 	"io/ioutil"
-	"net"
-	"os"
-	"path/filepath"
 	"sync"
 	"time"
 
-	"github.com/SkycoinProject/dmsg"
+	"github.com/SkycoinProject/skywire-mainnet/pkg/app/launcher"
```
Comment (nkryuchkov) [inline pkg/visor/config.go]: Could you please format imports?

File: pkg/visor/config.go [hunk 12]
```diff
@@ -47,258 +38,270 @@ var (
...
 
-	if c.Path == nil {
+	return c.flush()
+}
+
+func (c *Config) flush() error {
+	switch c.path {
+	case "":
 		return ErrNoConfigPath
+	case "STDIN":
```
Comment (nkryuchkov) [inline pkg/visor/config.go]: What do you think about creating a `const` for `"STDIN"`?

File: pkg/visor/config.go [hunk 13]
```diff
@@ -47,258 +38,270 @@ var (
...
+		return nil
+	default:
```
Comment (nkryuchkov) [inline pkg/visor/config.go]: We can remove this line.

File: pkg/visor/config.go [hunk 14]
```diff
@@ -5,39 +5,30 @@ import (
...
-	DefaultAppsPath = "./apps"
-	// DefaultLogLevel is used for default config generation and if it is not set in config.
-	DefaultLogLevel = "info"
-	// DefaultSTCPPort ???
-	// TODO: Define above or remove below.
-	DefaultSTCPPort = 7777
+
+	// ConfigVersion of the visor config.
+	// TODO: Put this in skyenv?
+	ConfigVersion = "v1.0.0"
```
Comment (nkryuchkov) [inline pkg/visor/config.go:30]: I think we need to decide if we are going to use it. If so, it's worth modifying when the config is changed.

File: pkg/visor/duration.go [hunk 15]
```diff
@@ -7,6 +7,7 @@ import (
 )
 
 // Duration wraps around time.Duration to allow parsing from and to JSON
+// TODO: Put this in a suitable module.
```
Comment (nkryuchkov) [inline pkg/visor/duration.go:10]: Do we need our custom type at all?

File: pkg/visor/init.go [hunk 16]
```diff
@@ -0,0 +1,432 @@
...
+
+	v.pushCloseStack("cli.listener", func() bool {
+		return report(cliL.Close())
+	})
+
+	rpcS, err := newRPCServer(v, "CLI")
+	if err != nil {
+		return report(fmt.Errorf("failed to start rpc server for cli: %w", err))
+	}
+	go rpcS.Accept(cliL) // We not not use sync.WaitGroup here as it will never return anyway.
```
Comment (nkryuchkov) [inline pkg/visor/init.go]: `We not not` looks like a typo

File: pkg/visor/init.go [hunk 17]
```diff
@@ -0,0 +1,432 @@
...
+
+	if conf == nil {
+		v.log.Info("'uptime_tracker' is not configured, skipping.")
+		return true
+	}
+
+	ut, err := utclient.NewHTTP(conf.Addr, v.conf.KeyPair.PubKey, v.conf.KeyPair.SecKey)
+	if err != nil {
+		// TODO(evanlinjin): We should design utclient to retry automatically instead of returning error.
+		//return report(err)
```
Comment (nkryuchkov) [inline pkg/visor/init.go:411]: Can this line be removed?

File: pkg/visor/init.go [hunk 18]
```diff
@@ -0,0 +1,432 @@
...
+		v.log.WithError(err).Warn("Failed to connect to uptime tracker.")
+		return true
+	}
+
+	log := v.MasterLogger().PackageLogger("uptime_tracker")
+	ticker := time.NewTicker(1 * time.Second)
```
Comment (nkryuchkov) [inline pkg/visor/init.go]: I'd extract `1 * time.Second` to a package-level constant.

File: pkg/visor/rpc.go [hunk 19]
```diff
@@ -9,6 +9,8 @@ import (
 	"os"
 	"time"
 
+	"github.com/SkycoinProject/skywire-mainnet/pkg/app/launcher"
```
Comment (nkryuchkov) [inline pkg/visor/rpc.go]: Not sorted.

File: pkg/visor/rpc_client.go [hunk 20]
```diff
@@ -11,6 +11,8 @@ import (
 	"sync"
 	"time"
 
+	"github.com/SkycoinProject/skywire-mainnet/pkg/app/launcher"
+
```
Comment (nkryuchkov) [inline pkg/visor/rpc_client.go]: Imports are not sorted, the empty line is redundant.

File: pkg/visor/visor.go [hunk 21]
```diff
@@ -2,53 +2,32 @@
...
-	"strconv"
 	"strings"
-	"sync"
 	"syscall"
 	"time"
 
-	"github.com/SkycoinProject/dmsg"
+	"github.com/sirupsen/logrus"
+
+	"github.com/SkycoinProject/skywire-mainnet/pkg/app/launcher"
```
Comment (nkryuchkov) [inline pkg/visor/visor.go]: Not sorted.

File: pkg/visor/visor.go [hunk 22]
```diff
@@ -60,619 +39,170 @@ const (
...
-	} else {
-		visor.logger.Info("Router closed cleanly.")
-	}
+	for i, ce := range v.closeStack {
 
-	return err
-}
+		start := time.Now()
+		done := make(chan bool, 1)
+		t := time.NewTimer(time.Second * 2)
```
Comment (nkryuchkov) [inline pkg/visor/visor.go]: Could you please extract a constant?

**Review by nkryuchkov — COMMENTED** [review]
Really great improvements! Good job!

File: pkg/visor/config.go (hunk 14 above)
Comment (evanlinjin) [reply pkg/visor/config.go:30]: @nkryuchkov I'm thinking of moving config into a separate module, maybe `/pkg/visor/visorconfig` and having versioned structs. E.g. `visorconfig.V1` or something.

File: pkg/visor/duration.go (hunk 15 above)
Comment (evanlinjin) [reply pkg/visor/duration.go:10]: @nkryuchkov I think it makes the config a bit cleaner. Other we will need to work with nanoseconds.

File: pkg/visor/config.go (hunk 14 above)
Comment (nkryuchkov) [reply pkg/visor/config.go:30]: @evanlinjin That sounds like a good idea.

File: pkg/visor/init.go (hunk 17 above)
Comment (evanlinjin) [reply pkg/visor/init.go:411]: @nkryuchkov I've left it there to give an example of how it should function once the above `TODO` comment is implemented. 😅

File: pkg/visor/init.go (hunk 18 above)
Comment (evanlinjin) [reply pkg/visor/init.go]: @nkryuchkov I've made it only as part of the function's scope as I feel like it makes more sense that way. Hope you are okay with this?

File: pkg/visor/rpc.go (hunk 19 above)
Comment (evanlinjin) [reply pkg/visor/rpc.go]: @nkryuchkov We need to find out whether we can make `goimports` do this automatically for us, or we need to write our own `go generate` tool.

File: pkg/visor/init.go (hunk 18 above)
Comment (nkryuchkov) [reply pkg/visor/init.go]: @evanlinjin Yes

File: pkg/visor/rpc.go (hunk 19 above)
Comment (nkryuchkov) [reply pkg/visor/rpc.go]: @evanlinjin Yeah, sorting imports is a pain 😔

Comment (nkryuchkov) [reply pkg/visor/rpc.go]: https://github.com/SkycoinProject/skywire-mainnet/issues/361

**Pull request merged**